syntax = "proto3";

// Shared website configuration
message WebConfig {
    message UserConfig {
        message HomePoint {
            float lon = 1;
            float lat = 2;
        }
        bool activity_tracking = 1;
        bool crash_reporting = 2;
        optional HomePoint home_point = 3;
    }
    string api_url = 1;
    float map_query_area_max_size = 2;
    float note_query_area_max_size = 3;
    optional UserConfig user_config = 4;
}

// Structure for transmitting user passwords
message TransmitUserPassword {
    optional string legacy = 1;
    optional bytes v1 = 2;
}

// Generic bounding box structure
message SharedBounds {
    float min_lon = 1;
    float min_lat = 2;
    float max_lon = 3;
    float max_lat = 4;
}

// Elements rendering structure
message RenderElementsData {
    message Node {
        uint64 id = 1;
        float lon = 2;
        float lat = 3;
    }
    message Way {
        uint64 id = 1;
        string line = 2;
        bool area = 3;
    }
    repeated Node nodes = 1;
    repeated Way ways = 2;
    optional bool too_much_data = 3;
}

// Changesets rendering structure
message RenderChangesetsData {
    message Changeset {
        message User {
            string name = 1;
            string avatar_url = 2;
        }
        uint64 id = 1;
        optional User user = 2;
        repeated SharedBounds bounds = 3;
        bool closed = 4;
        string timeago = 5;
        optional string comment = 6;
        uint64 num_comments = 7;
    }
    repeated Changeset changesets = 1;
}

// Notes rendering structure
message RenderNotesData {
    message Note {
        uint64 id = 1;
        float lon = 2;
        float lat = 3;
        string text = 4;
        bool open = 5;
    }
    repeated Note notes = 1;
}

// Map routing route result
message RoutingRoute {
    message Step {
        string line = 1;
        float distance = 2;  // in meters
        float time = 3;  // in seconds
        uint32 icon_num = 4;
        string text = 5;
    }
    message Elevation {
        float ascend = 1;  // in meters
        float descend = 2;  // in meters
    }
    string attribution = 1;
    repeated Step steps = 2;
    optional Elevation elevation = 3;
}

// Map routing resolve names result
message RoutingResolveNames {
    message Entry {
        string name = 1;
        SharedBounds bounds = 2;
        float lon = 3;
        float lat = 4;
    }
    optional Entry start = 1;
    optional Entry end = 2;
}

// Generic icon structure
message ElementIcon {
    string icon = 1;
    string title = 2;
}

// Partial changeset response params
message PartialChangesetParams {
    message Element {
        uint64 id = 1;
        uint64 version = 2;
        bool visible = 3;
        optional string name = 4;
        optional ElementIcon icon = 5;
    }
    uint64 id = 1;
    repeated SharedBounds bounds = 2;
    repeated Element nodes = 3;
    repeated Element ways = 4;
    repeated Element relations = 5;
}

// Partial element response params
message PartialElementParams {
    message Entry {
        string type = 1;
        uint64 id = 2;
        optional string role = 3;
        optional string name = 4;
        optional ElementIcon icon = 5;
    }
    string type = 1;
    repeated Entry members = 2;
    repeated Entry parents = 3;
    RenderElementsData render = 4;
}

// Partial note response params
message PartialNoteParams {
    uint64 id = 1;
    float lon = 2;
    float lat = 3;
    bool open = 4;
}

// Partial query features response params
message PartialQueryFeaturesParams {
    repeated RenderElementsData renders = 1;
}

// Partial search response params
message PartialSearchParams {
    optional string bounds_str = 1;
    repeated RenderElementsData renders = 2;
    bool where_is_this = 3;
}
